server/channels/app/export.go

```diff
@@ -194,6 +215,86 @@ func (a *App) exportVersion(writer io.Writer) *model.AppError {
 	return a.exportWriteLine(writer, versionLine)
 }
 
+func (a *App) exportRoles(ctx request.CTX, job *model.Job, writer io.Writer, schemeRoles map[string]bool) *model.AppError {
+	roles, appErr := a.Srv().Store().Role().GetAll()
+	if appErr != nil {
+		return model.NewAppError("exportSchemes", "app.role.get_all.app_error", nil, "", http.StatusInternalServerError).Wrap(appErr)
+	}
+
+	for _, role := range roles {
+		// We skip any roles that will be included as part of custom schemes.
+		if !schemeRoles[role.Name] {
+			if err := a.exportWriteLine(writer, ImportLineFromRole(role)); err != nil {
+				return err
+			}
+		}
+	}
+
+	updateJobProgress(ctx.Logger(), a.Srv().Store(), job, "roles_exported", len(roles))
+
+	return nil
+}
+
+func (a *App) exportSchemes(ctx request.CTX, job *model.Job, writer io.Writer, scope string, schemeRolesMap map[string]bool) *model.AppError {
+	var cnt int
+	pageSize := 100
+
+	roles, appErr := a.Srv().Store().Role().GetAll()
+	if appErr != nil {
+		return model.NewAppError("exportSchemes", "app.role.get_all.app_error", nil, "", http.StatusInternalServerError).Wrap(appErr)
+	}
+
+	rolesMap := make(map[string]*model.Role, len(roles))
+	for _, role := range roles {
+		rolesMap[role.Name] = role
+	}
+
+	for {
+		schemes, appErr := a.Srv().Store().Scheme().GetAllPage(scope, cnt, pageSize)
+		if appErr != nil {
```

Comment:
Couple of things:
1. This does not return an appErr. Just a normal error.
2. This will return a `ErrNoRows` if pageSize is a multiple of total number of rows. We need to distinguish that case and handle it.

Reply:
Thanks, on 2. that store method doesn't use `ErrNoRows`, and we are breaking the loop upon the returned slice's length being less than page size so I am thinking the only occurrence would be if there are no schemas, but again we are not checking for no rows returned.

Reply:
@agnivade I am thinking `sqlx.Select()` doesn't return `ErrNoRows` as it allows for returning empty slices. I've tested both with no schemes and page size 1 which should trigger the error if that wasn't the case.

Reply:
Ahhh indeed. Thanks, that's what I was missing.